Add lastUpdated property to Form

Chartman123 opened the pull request:
This fixes #1462 

Signed-off-by: Christian Hartmann <chris-hartmann@gmx.de>

Chartman123 commented:
@jotoeri could you please point me in the right direction, why my setLastUpdatedTimestamp function in the forms service doesn't work?

Chartman123 commented:
> And you first need to store the form entity you get from findById. wink

🤦🏻 



> $this->formMapper->update($form);

Thanks, this was the missing part.

Chartman123 commented:
@jotoeri The backend stuff should be ready now. However, I'm not sure how to proceed with the frontend, so that a changed form for example will also result in a reordered forms list in the AppNavigation. Shall we split this into a second PR?

jotoeri commented:
Hmm, to avoid confusion, i think we should shift it up, once updated/edited. That will of course not cover submissions, but will make clear the principle.
- Should suffice to just use the timestamp on the Frontend within a session, no need to have the exact timestamp of the backend.
- To not pass the event everywhere through our components you can also use the nextcloud/event-bus package.

And i wouldn't do it separately. Its closely connected.

Chartman123 commented:
@jotoeri It's already working for updated form properties. However, I still get some errors when updating question properties and question options. Perhaps you could already have a look at it... I will only be able to get back to work tomorrow in the evening.

susnux commented:
> I still get some errors when updating question properties and question options.

Do you still get those errors? For me it works perfectly when changing question properties like *required* or *shuffle options*.

Chartman123 commented:
@susnux I figured out the question related problems... now just the options/answers are missing.

I'm a little clueless what is triggered by the $emit('update:options',...) in dropdown and multiple questions. There is no @update:options in any of the vue components...

susnux commented:
> I'm a little clueless what is triggered by the $emit('update:options',...) in dropdown and multiple questions. There is no @update:options in any of the vue components...

I think it is used in `views/Create.vue`: `v-bind.sync="form.questions[index]"`. As this should bind `:options.sync`.

Chartman123 commented:
@susnux @jotoeri I've now added the re-sorting for changed options. My last problem is that this also moves the form up just on clicking inside the "add new option" input box. At this time, the options haven't change in the database, so basically the sorting is "wrong".

Chartman123 commented:
@susnux Your new ApiControllerTest is killing me... 😆 Could you please help me fix the failing checks here?

susnux commented:
You need to mock the `time()` function as the controller sets the `lastUpdated` value based on it. (see my commit)

Chartman123 commented:
Thank you! Now I can polish the integration tests and then everything should be fine :)

susnux commented on src/Forms.vue:
```diff
@@ -224,6 +225,14 @@ export default {
 	},
 
+	mounted() {
+		subscribe('change-last-updated', (id) => this.onLastUpdatedByEventBus(id))
```
I think for custom events we should stick to the event-bus naming scheme: `app-id:object:verb` e.g. `forms:last-updated:changed` or similar. (also to prevent naming clashes)

Chartman123 replied:
Done :)

jotoeri reviewed (approved):
So - now got into this one. Two small comments, maybe also squash a bit, but apart from that, it should be fine by me.

Chartman123 commented:
> So - now got into this one. Two small comments, maybe also squash a bit, but apart from that, it should be fine by me.

Fixed them and squashed the commits. I think it's ready to go now :)

jancborchardt reviewed (commented):
Good stuff, thank you @Chartman123! :) (Can't review the code, so only leaving comment instead of approval.)

susnux reviewed (approved):
:+1:

susnux commented:
@Chartman123 
I tested this using mariaDB, so I did not noticed it, but this breaks updating a server that runs on SQLite.
The `comment` option on the new database column does not work with SQLite:
https://www.doctrine-project.org/projects/doctrine-dbal/en/current/reference/schema-representation.html#common-options

```
In ExceptionConverter.php line 82:
                                                                                                     
  An exception occurred while executing a query: SQLSTATE[HY000]: General error: 1 incomplete input  
                                                                                                     

In Exception.php line 30:
                                                      
  SQLSTATE[HY000]: General error: 1 incomplete input  
                                                      

In Connection.php line 72:
                                                      
  SQLSTATE[HY000]: General error: 1 incomplete input  
                                                      
```

I do not know why a clean install work (no forms installed before), it only fails if a partial migration is required.

Chartman123 commented:
Strange, I copied the code from some older migration... 🙈
